chainer/training/extensions/evaluator.py

```diff
@@ -0,0 +1,168 @@
+import copy
+
+import six
+
+from chainer.dataset import convert
+from chainer.dataset import iterator as iterator_module
+from chainer import link
+from chainer import reporter as reporter_module
+from chainer.training import extension
+from chainer import variable
+
+
+class Evaluator(extension.Extension):
+
+    """Trainer extension to evaluate models on a validation set.
+
+    This extension evaluates the current models by a given evaluation function.
+    It creates a :class:`~chainer.Reporter` object to store values observed in
+    the evaluation function on each iteration. The report for all iterations
+    are aggregated to :class:`~chainer.DictSummary`. The collected mean values
+    are further reported to the reporter object of the trainer, where the name
+    of each observation is prefixed by the evaluator name. It also report their
+    standard deviations, where the names are also suffixed by ``.std``.
+
+    Evaluator has a structure to customize similar to that of
+    :class:`~chainer.training.StandardUpdater`. The main differences are:
+
+    - There are no optimizers in an evaluator. Instead, it holds links
+      to evaluate.
+    - An evaluation loop function is used instead of an update function.
+    - Preparation routine can be customized, which is called before each
+      evaluation. It can be used, e.g., to initialize the state of stateful
+      recurrent networks.
+
+    There are two ways to modify the evaluation behavior besides setting a
+    custom evaluation function. One is by setting a custom evaluation loop via
+    the ``eval_loop`` argument. The other is by inheriting this class and
+    overriding the :meth:`evaluate` method. In latter case, users have to
+    create and handle a reporter object manually. Users also have to copy the
+    iterators before using them, in order to reuse them at the next time of
+    evaluation.
+
+    This extension is called at the end of each epoch by default.
+
+    Args:
+        iterator: Dataset iterator for the validation dataset. It can also be
+            a dictionary of iterators. If this is just an iterator, the
+            iterator is registered by the name ``'main'``.
+        target: Link object or a dictionary of links to evaluate. If this is
+            just a link object, the link is registered by the name ``'main'``.
+        eval_loop: Evaluation loop. This is a function that takes the evaluator
+            object as the argument, and returns a dictionary of statistics for
+            observed values. The default routine uses ``converter`` and
```

Comment:
"If it is omitted, default routine is made from converter and eval_func if specified."